### ISSUE TITLE
evaluate with usb camera not worked

### ISSUE THREAD

**Issue by ntut108318099** [haotian-liu/yolact_edge#17]
Hello,thanks for your research first ~
I am the student in graduate school now. I am very interested in this paper.

I would like to use usb camera to evaluate model,so I ran the following command depends on your Readme.md file.

```
python eval.py --coco_transfer --trained_model=weights/yolact_edge_54_800000.pth --score_threshold=0.3 --top_k=100 --video_multiframe=2 --video=0
```
But after I ran the command,I found the model read the json file and start calibration dataset for backbone of 100 images,it seems not evaluate with camera. (Like the following picture)
![image](https://user-images.githubusercontent.com/56862464/103395959-a2368d00-4b6b-11eb-8fc7-1b43743b16bb.png)

Was the result progress correct ?   Could you tell me what's probably going on?
Thanks you, Have a nice day,and Happy new year!!

(P.S. Why use "--coco_transfer " ? what did the option mean? thanks for your help!!)

**Comment by rrivera1849** [comment]
Hi there! Did the code throw an error or did it just seem stuck in converting to TensorRT? Note that if this is the first time converting the model to TensorRT the calibration will take a long time. For example, on a Jetson AGX Xavier it took me 25 minutes to convert the entire model to TensorRT. On my side the code finished calibrating but then I ran into another issue when evaluating with a webcam. You'll want to run the following command:

```python
python eval.py --coco_transfer --trained_model=weights/yolact_edge_54_800000.pth --score_threshold=0.3 --top_k=100 --video_multiframe=1 --video=0
``` 

The "video_multiframe" argument had to be changed to "1". This is because when we convert to TensorRT we're using batch sizes of one. We will add in an option that does the TRT conversion with higher batch sizes in order to support higher batch sizes during evaluation on videos / webcams. We will also change the README so that others don't get confused. 

The "--coco_transfer" flag is used to make models that were trained with YOLACT compatible with YolactEdge.

**Comment by rrivera1849** [comment]
The README and code has been updated to support higher batch sizes than 1 with TRT. The following command should work:

```python
python eval.py --coco_transfer --trained_model=weights/yolact_edge_54_800000.pth --score_threshold=0.3 --top_k=100 --video_multiframe=2 --trt_batch_size 2 --video=my_video.mp4
```

**Comment by ntut108318099** [comment]
Thanks for your reply,I will try it immediately!!

Have a nice day ~~

**Comment by ntut108318099** [comment]
Sorry when I ran the command,the command show some "UserWarning" after loading the model,could you tell me what this warning mean to be ?

![image](https://user-images.githubusercontent.com/56862464/103505244-bf819900-4e94-11eb-991e-643ca83410d2.png)

And It stuck on converting to TensorRT about 5 minutes, It showed killed. 
I think the setup is not properly..

I use anaconda environment to build the code,did it probably cause the problem?
Here`s my environment setup , please take a look at~~ 
Thanks a lot.

```
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       1_gnu    conda-forge
absl-py                   0.11.0                   pypi_0    pypi
appdirs                   1.4.4                    pypi_0    pypi
ca-certificates           2020.12.5            ha878542_0    conda-forge
cachetools                4.2.0                    pypi_0    pypi
certifi                   2020.12.5        py37h89c1867_0    conda-forge
chardet                   4.0.0                    pypi_0    pypi
cudatoolkit               10.2.89              h8f6ccaa_6    conda-forge
cudnn                     8.0.0                cuda10.2_0    nvidia
cycler                    0.10.0                   pypi_0    pypi
cython                    0.29.21                  pypi_0    pypi
decorator                 4.4.2                    pypi_0    pypi
future                    0.18.2                   pypi_0    pypi
gitdb                     4.0.5                    pypi_0    pypi
gitpython                 3.1.11                   pypi_0    pypi
google-auth               1.24.0                   pypi_0    pypi
google-auth-oauthlib      0.4.2                    pypi_0    pypi
graphsurgeon              0.4.5                    pypi_0    pypi
grpcio                    1.34.0                   pypi_0    pypi
idna                      2.10                     pypi_0    pypi
importlib-metadata        3.3.0                    pypi_0    pypi
kiwisolver                1.3.1                    pypi_0    pypi
ld_impl_linux-64          2.35.1               hea4e1c9_1    conda-forge
libblas                   3.9.0                6_openblas    conda-forge
libcblas                  3.9.0                6_openblas    conda-forge
libffi                    3.3                  h58526e2_2    conda-forge
libgcc-ng                 9.3.0               h5dbcf3e_17    conda-forge
libgfortran-ng            9.3.0               he4bcb1c_17    conda-forge
libgfortran5              9.3.0               he4bcb1c_17    conda-forge
libgomp                   9.3.0               h5dbcf3e_17    conda-forge
liblapack                 3.9.0                6_openblas    conda-forge
libopenblas               0.3.12          pthreads_h4812303_1    conda-forge
libstdcxx-ng              9.3.0               h2ae2ef3_17    conda-forge
mako                      1.1.3                    pypi_0    pypi
markdown                  3.3.3                    pypi_0    pypi
markupsafe                1.1.1                    pypi_0    pypi
matplotlib                3.3.3                    pypi_0    pypi
ncurses                   6.2                  h58526e2_4    conda-forge
numpy                     1.19.4           py37haa41c4c_2    conda-forge
oauthlib                  3.1.0                    pypi_0    pypi
opencv-python             4.5.1.48                 pypi_0    pypi
openssl                   1.1.1i               h7f98852_0    conda-forge
pillow                    8.1.0                    pypi_0    pypi
pip                       20.3.3             pyhd8ed1ab_0    conda-forge
protobuf                  3.14.0                   pypi_0    pypi
pyasn1                    0.4.8                    pypi_0    pypi
pyasn1-modules            0.2.8                    pypi_0    pypi
pycocotools               2.0                      pypi_0    pypi
pycuda                    2020.1                   pypi_0    pypi
pyparsing                 2.4.7                    pypi_0    pypi
python                    3.7.9           hffdb5ce_0_cpython    conda-forge
python-dateutil           2.8.1                    pypi_0    pypi
python_abi                3.7                     1_cp37m    conda-forge
pytools                   2020.4.4                 pypi_0    pypi
readline                  8.0                  he28a2e2_2    conda-forge
requests                  2.25.1                   pypi_0    pypi
requests-oauthlib         1.3.0                    pypi_0    pypi
rsa                       4.6                      pypi_0    pypi
setuptools                49.6.0           py37he5f6b98_2    conda-forge
six                       1.15.0                   pypi_0    pypi
smmap                     3.0.4                    pypi_0    pypi
sqlite                    3.34.0               h74cdb3f_0    conda-forge
tensorboard               2.4.0                    pypi_0    pypi
tensorboard-plugin-wit    1.7.0                    pypi_0    pypi
tensorrt                  7.1.3.4                  pypi_0    pypi
termcolor                 1.1.0                    pypi_0    pypi
tk                        8.6.10               h21135ba_1    conda-forge
torch                     1.6.0                    pypi_0    pypi
torchvision               0.7.0                    pypi_0    pypi
typing-extensions         3.7.4.3                  pypi_0    pypi
uff                       0.6.9                    pypi_0    pypi
urllib3                   1.26.2                   pypi_0    pypi
werkzeug                  1.0.1                    pypi_0    pypi
wheel                     0.36.2             pyhd3deb0d_0    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
zipp                      3.4.0                    pypi_0    pypi
zlib                      1.2.11            h516909a_1010    conda-forge
```

**Comment by ntut108318099** [comment]
Sorry to bother you, @haotian-liu  I run the code successful without conda environment!
![image](https://user-images.githubusercontent.com/56862464/103645015-4e281000-4f92-11eb-8b03-670ea72f5053.png)

But, Here's some question I want to ask : 
1. The command called train_model : `yolact_edge_54_800000.pth ` to evaluate
 Can I called the  trt model ( convert result) to evaluate? 

2. I want to do transfer learning on COCO person dataset(means only 1 person label), what can I do?
Can I just used the method mentioned by YOLACT (https://github.com/dbolya/yolact/issues/36) 
and set 'start_iter=0' to train model?

3.After that,I will try to compile the code with ROS wrapper,just like 'yolact_ros'(https://github.com/Eruvae/yolact_ros),Did you have any idea to compile with ROS?

I will try my best to compile with ROS. If you have interested,I can share the code .

Thanks for your supply~~

**Comment by haotian-liu** [comment]
Cool and thank you for your interests. For your questions:

1. After the model parts is compiled to TensorRT (backbone, protonet, etc.), the network will store the cache alongside with the model. So you can call the trt model (cache) to evaluate. However, just to note that this cache is mainly used to save time of re-evaluation (so that we avoid conversion of backbone over and over again), some modules are not stored as the cache and are compiled in every runs. Besides, another information I could provide is that the TensorRT converted `.trt` models seems device- (or at least GPU-) dependent, and when I tried to inference on Jetson with the TensorRT model converted on RTX 2080 Ti, it will raise an error.
2. Yes you should be able to do the same for fine-tuning as YOLACT.
3. Since I am not familiar with ROS, I am unable to provide as much information on this.

**Comment by ntut108318099** [comment]
Thanks for your quick reply, please let me ask you another question 😅 

If I want to catch the mask information directly , where can I find in eval.py ?

After I research the code,I think the following part is about mask information,but I have no idea how to get the mask information only. 

The thing I want to do is get the mask information ,so I can use opencv to draw it with only black image (cv::Mat(zeros))

```sh
 First, draw the masks on the GPU where we can do it really fast

    # Beware: very fast but possibly unintelligible mask-drawing code ahead
    # I wish I had access to OpenGL or Vulkan but alas, I guess Pytorch tensor operations will have to suffice
    if args.display_masks and cfg.eval_mask_branch:
        # After this, mask is of size [num_dets, h, w, 1]
        masks = masks[:num_dets_to_consider, :, :, None]
        # Prepare the RGB images for each mask given their color (size [num_dets, h, w, 1])
        colors = torch.cat([get_color(j, on_gpu=img_gpu.device.index).view(1, 1, 1, 3) for j in range(num_dets_to_consider)], dim=0)
        masks_color = masks.repeat(1, 1, 1, 3) * colors * mask_alpha
        # This is 1 everywhere except for 1-mask_alpha where the mask is
        inv_alph_masks = masks * (-mask_alpha) + 1
        # I did the math for this on pen and paper. This whole block should be equivalent to:
        #    for j in range(num_dets_to_consider):
        #        img_gpu = img_gpu * inv_alph_masks[j] + masks_color[j]
        masks_color_summand = masks_color[0]
        if num_dets_to_consider > 1:
            inv_alph_cumul = inv_alph_masks[:(num_dets_to_consider-1)].cumprod(dim=0)
            masks_color_cumul = masks_color[1:] * inv_alph_cumul
            masks_color_summand += masks_color_cumul.sum(dim=0)
        img_gpu = img_gpu * inv_alph_masks.prod(dim=0) + masks_color_summand
```

**Comment by haotian-liu** [comment]
Please refer to this block of [code](https://github.com/haotian-liu/yolact_edge/blob/master/eval.py#L160-L170) where you can get mask information.
